### PR TITLE
[Utility] Foundational bugs, tests, code cleanup and improvements (1 / 2)

### LINKED ISSUE #475: [Utility] Utility Module Foundation - Part 1 / 2
## Objective

Partial refactor, cleanup, and code improvement of the utility module.

## Origin Document

Originally intended as follow-up work for #325.

When work on #443 started, while also looking at #325, it became evident that the logic in the utility module is hard to follow and understand, while also being filled with minor hard-to-find bugs.

## Goals

- **Unblock development**: Create a foundation on which we design utility protocols and implement new features
- **Reduce cognitive overhead**: Improve code readability and navigation for both internal and external contributors
- **Enable onboarding other engineers**: Clarify both source code and tests through comment and source code improvements

## Deliverable

- [ ] A PR that updates a significant portion (1/N) of the utility module
- [ ] Documentation & explanation of the changes that were made and next steps

## Non-goals / Non-deliverables

- Updating the entire module in a single PR (PR would be too large)
- Adding new features to the utility module
- Increasing/improving the test coverage to the po

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @Olshansk

### PR DESCRIPTION
**Pull request by Olshansk**
## Description

This is the first of several changes necessary to refactor and improve the utility module to enable implementation of future, more complex, interfaces.

## Issue

Fixes #475

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [x] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other 

### Changes

**Utility**
- Add a `Validatable` type for basic validation
- Split business logic specific to certain actors (e.g. validator reward, app relays, message handling) into separate files
- Reduce the scope of functions and types that shouldn’t be exposed
- Upgraded the actors tests - a lot went here to help with understanding what’s going on but it’s still just a start
- Remove the `Context` struct; unnecessary abstraction
- Added comments and guidance on message, transaction and signature validation
- Added `ITransaction`, an interface for the `Transaction` protocol to help capture the functionality it adds to the core type

**Code Confusion**
- Remove `IUnstakingActor` and use `UnstakingActor` directly. Guideline for removing future unnecessary types (e.g. TxResult)
- Delineate between unstaking & unbonding in a few places throughout the codebase

**Bugs**
- Avoid unstaking all actors when intending to UnstakeMsg paused actors only (major bug in persistence sql query)
- `tx.Equals` was comparing the same transaction against itself (major bug)
- Staking status enums in utility did not reflect the same protocol as in persistence (needs to be consolidated)

**Code optimization**
- Avoid redundant `byte` <-> `string` conversions in several places
- Avoid redundant `bigInt` <-> `string` conversions in several places

**Code health**
- Add comments explaining the use/responsibility of various types
- Consolidate BigInt/String converters used in other parts of the codebase
- Consolidate some types used everywhere (e.g. `actorTypes`)
- Reduce the code footprint of the `codec` package & add some TODOs
- Removed unused double sign code (moved to #432 for reference)

### Files focused on

utility
- [x] ├── account.go
- [x] ├── account_test.go
- [x] ├── actor.go
- [x] ├── actor_test.go
- [ ] ├── application.go
- [ ] ├── application_test.go
- [ ] ├── block.go
- [ ] ├── block_test.go
- [ ] ├── context.go
- [ ] ├── doc
- [ ] │   ├── CHANGELOG.md
- [ ] │   ├── PROTOCOL_RELAY.md
- [ ] │   ├── PROTOCOL_SESSION.md
- [ ] │   └── README.md
- [ ] ├── gov.go
- [ ] ├── gov_test.go
- [ ] ├── message_handler.go
- [ ] ├── message_test.go
- [ ] ├── module.go
- [ ] ├── module_test.go
- [ ] ├── service
- [ ] │   └── service.go
- [ ] ├── session.go
- [ ] ├── transaction.go
- [ ] ├── transaction_test.go
- [ ] ├── types
- [x] │   ├── constants.go
- [x] │   ├── error.go
- [x] │   ├── gov.go
- [x] │   ├── message.go
- [x] │   ├── message_staking.go
- [x] │   ├── message_test.go
- [ ] │   ├── proto
- [ ] │   │   ├── message.proto
- [ ] │   │   ├── stake_status.proto
- [ ] │   │   ├── transaction.proto
- [ ] │   │   ├── tx_result.proto
- [ ] │   │   └── utility_messages.proto
- [x] │   ├── relay_chain.go
- [x] │   ├── relay_chain_test.go
- [x] │   ├── signature.go
- [ ] │   ├── transaction.go
- [ ] │   ├── transaction_test.go
- [x] │   ├── tx_fifo_mempool.go
- [x] │   ├── tx_fifo_mempool_test.go
- [ ] │   ├── tx_result.go
- [x] │   └── validatable.go
- [ ] └── validator.go

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [x] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by Olshansk** [comment]
I left many TODOs & comments through the code, but the following is also just a list of idea/notes I made while working on this PR. It SHOULD NOT be taken verbatim but just as a direction of other (minor & major) work we'll need to do.

**Tickets to create:**
- [ ] Rename `ServiceNode` to `Servicer` everywhere
- [ ] Rename “State Sync” to “Block Sync” where appropriate
- [ ] Need to understand & re-evalute the definition and usage of `TxResult` and `TransactionResult`
- [ ] Need to create E2E & Unit Tests for:
    - [ ] Burning jailed protocol actors
    - [ ] Burning double signing validators
    - [ ] Validating that stake from paused, unstaking and unbinding actors is handled correctly
    - [ ] Application relay quota is properly computed
    - [ ] Validating the actor stake goes to/from the appropriate pool when staking/unstaking
- [ ] Consolidate the enum & Consolidate actor staking status
- [ ] Remove all instances of `[]byte` for addresses and use `string` instead
- [ ] Changing all heights to use uint64

**Topics to discuss:**
- Updating shared module interfaces with semantic types (i.e. protobufs) instead of lengthy basic types (e.g. replace `[]byte` with `proto.Transaction`)
- Removing or consolidating Pocket’s custom `Error` in utility across the entire codebase
- Discuss and explain the difference between unstaking and unbinding
- Should amounts be strings?
- What is a Tx?
    - Tl;dr Message(s) with a signature and Nonce

File: persistence/block.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 	coreTypes "github.com/pokt-network/pocket/shared/core/types"
 )
 
+// INCOMPLETE: Need to think through how we will handle pruned vs non-pruned txIndexers
```
Comment (deblasis) [inline persistence/block.go:15]: I think that there's a lot packed into this comment, shall we create an issue to describe what you mean in detail and tag it here?

File: shared/codec/codec.go [hunk 2]
```diff
@@ -5,56 +5,45 @@ import (
...
+type Codec interface {
 	Marshal(proto.Message) ([]byte, error)
 	Unmarshal([]byte, proto.Message) error
 	ToAny(proto.Message) (*anypb.Any, error)
 	FromAny(*anypb.Any) (proto.Message, error)
+	Clone(proto.Message) proto.Message
 }
 
-var _ Codec = &ProtoCodec{}
+type ICodec proto.Message
```
Comment (deblasis) [inline shared/codec/codec.go:18]: I don't like this TBH. Care to elaborate the reasoning behind the change here?

File: utility/types/message.go [hunk 3]
```diff
@@ -1,53 +1,25 @@
...
-	PortOutOfRange   = "invalid port, out of valid port range"
-	NoPeriod         = "must contain one '.'"
-	MaxPort          = 65535
 )
 
+// A message is a component of a transaction (excluding metadata such as the signature)
+// defining the action driving the state transition
 type Message interface {
-	proto.Message
+	codec.ICodec
```
Comment (deblasis) [inline utility/types/message.go:15]: I don't like this TBH. Care to elaborate the reasoning behind the change here?

File: shared/core/types/proto/actor.proto [hunk 4]
```diff
@@ -7,8 +7,8 @@ option go_package = "github.com/pokt-network/pocket/shared/core/types";
 enum ActorType {
   ACTOR_TYPE_UNSPECIFIED = 0;
   ACTOR_TYPE_APP = 1;
-  ACTOR_TYPE_SERVICENODE = 2;// TODO: Consider renaming to `servicers` throughout the codebase.
-  ACTOR_TYPE_FISH = 3;
+  ACTOR_TYPE_SERVICENODE = 2;
+  ACTOR_TYPE_FISH = 3;// TODO: Consider renaming to `servicers` throughout the codebase.
```
Comment (deblasis) [inline shared/core/types/proto/actor.proto:11]: Oh really? Fishermen to be renamed servicers?

File: shared/core/types/actor.go [hunk 5]
```diff
@@ -1,5 +1,15 @@
...
+// REFACTOR: Evaluate a way to define an interface for the actors which would enable to easily
+//           remove all the `switch actorType` statements in the codebase.
+
+var ActorTypes = []ActorType{
+	ActorType_ACTOR_TYPE_APP,
+	ActorType_ACTOR_TYPE_FISH,
+	ActorType_ACTOR_TYPE_SERVICENODE,
+	ActorType_ACTOR_TYPE_VAL,
+}
+
```
Comment (deblasis) [inline shared/core/types/actor.go:12]: I am not sure this is needed. We already have the generated:

```go
// Enum value maps for ActorType.
var (
	ActorType_name = map[int32]string{
		0: "ACTOR_TYPE_UNSPECIFIED",
		1: "ACTOR_TYPE_APP",
		2: "ACTOR_TYPE_SERVICENODE",
		3: "ACTOR_TYPE_FISH",
		4: "ACTOR_TYPE_VAL",
	}
	ActorType_value = map[string]int32{
		"ACTOR_TYPE_UNSPECIFIED": 0,
		"ACTOR_TYPE_APP":         1,
		"ACTOR_TYPE_SERVICENODE": 2,
		"ACTOR_TYPE_FISH":        3,
		"ACTOR_TYPE_VAL":         4,
	}
)
```

in `actor.pb.go` and I think we should rely on them instead. This way it's future proof as well since changes in the enum woudn't require changes here if that makes sense.

File: shared/crypto/sha3.go [hunk 6]
```diff
@@ -17,9 +17,7 @@ func SHA3Hash(bz []byte) []byte {
 	return hasher.Sum(nil)
 }
 
-// GetHashStringFromBytes returns the hexadecimal encoding in string format of the SHA3Hash of the bytes passed in as argument
-//
-// Typically used to compute a TransactionHash
+// Returns hex(SHA3Hash(bytesArgument)); typically used to compute a TransactionHash
 func GetHashStringFromBytes(bytes []byte) string {
```
Comment (deblasis) [inline shared/crypto/sha3.go:21]: By convention, comments that describe functions start with the name of the function. If I recall correctly it's because of how godoc works

File: shared/modules/types/proto/unstaking_actor.proto [hunk 7]
```diff
@@ -0,0 +1,11 @@
+syntax = "proto3";
+
+package modules;
+
+option go_package = "github.com/pokt-network/pocket/shared/modules/types";
+
+message UnstakingActor {
```
Comment (deblasis) [inline shared/modules/types/proto/unstaking_actor.proto:7]: Not actionable, just TOL. I wonder if this should be either part of `coreTypes`

File: shared/modules/types/README.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# Modules Types
+
+The types defined in this package are used for cross-module communication.
+
+The order of serialization is less critical, compared to the core types, but defines the "contract" of what one module expects of another.
```
Comment (deblasis) [inline shared/modules/types/README.md:5]: I am not 100% sold on this distinction but I can close an eye on it I guess

File: utility/types/message_staking.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	httpsPrefix      = "https://"
+	httpPrefix       = "http://"
+	colon            = ":"
+	period           = "."
+	invalidURLPrefix = "the url must start with http:// or https://"
+	portRequired     = "a port is required"
+	nonNumberPort    = "invalid port, cant convert to integer"
+	portOutOfRange   = "invalid port, out of valid port range"
+	noPeriod         = "must contain one '.'"
+	maxPort          = 65535
```
Comment (deblasis) [inline utility/types/message_staking.go:24]: Not sure this file is the best spot for these. Some of them are quite general purpose

File: utility/types/message_test.go [hunk 10]
```diff
@@ -50,66 +51,6 @@ func TestMessage_ChangeParameter_ValidateBasic(t *testing.T) {
 	require.Equal(t, ErrEmptyParamValue().Code(), msgMissingParamValue.ValidateBasic().Code())
 }
 
-func TestMessage_DoubleSign_ValidateBasic(t *testing.T) {
```
Comment (deblasis) [inline utility/types/message_test.go:53]: Did we lose this test?

File: utility/application.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+	"math/big"
+
+	"github.com/pokt-network/pocket/shared/converters"
+	typesUtil "github.com/pokt-network/pocket/utility/types"
+)
+
+var (
+	// TECHDEBT: Re-evalute the denomination of tokens used throughout the codebase. `MillionInt` is
+	// currently used to convert POKT to uPOKT but this is not clear throughout the codebase.
+	MillionInt = big.NewFloat(1000000)
```
Comment (deblasis) [inline utility/application.go:14]:
```suggestion
	MillionInt = big.NewFloat(1e6)
```

File: utility/block.go [hunk 12]
```diff
@@ -21,25 +24,26 @@ import (
...
-	mempool := u.GetBus().GetUtilityModule().GetMempool()
+
+	mempool := u.getBus().GetUtilityModule().GetMempool()
 	for !mempool.IsEmpty() {
 		txBytes, err := mempool.PopTx()
 		if err != nil {
 			return "", nil, err
 		}
-		transaction, err := typesUtil.TransactionFromBytes(txBytes)
+		transaction, err := typesUtil.TxFromBytes(txBytes)
```
Comment (deblasis) [inline utility/block.go:46]: Not a strong opinion: shouldn't this be `tx` now? Actually I'd consider renaming the type as well from `Transaction` to `Tx`. I know it's interchangeable but since we are touching it, let's give it a little bit more love?

File: utility/block.go [hunk 13]
```diff
@@ -53,28 +57,28 @@ func (u *UtilityContext) CreateAndApplyProposalBlock(proposer []byte, maxTransac
 			}
 			break // we've reached our max
 		}
-		txResult, err := u.ApplyTransaction(txIndex, transaction)
+		txResult, err := u.applyTransaction(txIndex, transaction)
```
Comment (deblasis) [inline utility/block.go:60]: Similar to a previous comment. There's inconsistency here now. I'd expect to read `u.applyTx(txIndex, tx)` here. And the 2inch dusty fan inside my brain is spinning faster :)

File: utility/block.go [hunk 14]
```diff
@@ -110,13 +114,13 @@ func (u *UtilityContext) ApplyBlock() (string, error) {
...
 		// Validate and apply the transaction to the Postgres database
-		txResult, err := u.ApplyTransaction(index, tx)
+		txResult, err := u.applyTransaction(index, tx)
 		if err != nil {
 			return "", err
 		}
 
-		if err := u.Context.IndexTransaction(txResult); err != nil {
-			u.logger.Fatal().Err(err).Msg("TODO(#327): We can apply the transaction but not index it. Crash the process for now")
+		if err := u.persistenceContext.IndexTransaction(txResult); err != nil {
```
Comment (deblasis) [inline utility/block.go:122]: Here too `IndexTransaction` and not `IndexTx`? It makes me wonder if it's the same thing (If I use a bit of empathy, like if I am reading this for the first time)

File: utility/context.go [hunk 15]
```diff
@@ -3,147 +3,124 @@ package utility
...
-	Context *Context // IMPROVE: Rename to `persistenceContext` or `storeContext` or `reversibleContext`?
+type utilityContext struct {
+	bus    modules.Bus
+	height int64
 
-	// Data related to the Block being proposed
-	// TECHDEBT: When we consolidate everything to have a single `Block` object (a struct backed by a protobuf),
-	//           this can be simplified to just point to that object.
+	persistenceContext modules.PersistenceRWContext
+	// INVESTIGATE: Can we use the
```
Comment (deblasis) [inline utility/context.go:15]: Can we use the.... 
GeenoBttfDeloreanFluxcompensatorGIF

Please tell me that it's the flux capacitor, please please please!

Narrator: yes, he is old

File: utility/context.go [hunk 16]
```diff
@@ -3,147 +3,124 @@ package utility
...
 	return nil
 }
 
-func (u *UtilityContext) GetBus() modules.Bus {
+func (u *utilityContext) getBus() modules.Bus {
 	return u.bus
 }
 
-func (u *UtilityContext) WithBus(bus modules.Bus) *UtilityContext {
+func (u *utilityContext) withBus(bus modules.Bus) *utilityContext {
```
Comment (deblasis) [inline utility/context.go:117]: oh I must have missed these guys. `withBus`? why not `setBus` since we are here?

**Review by deblasis — CHANGES_REQUESTED** [review]
Solid piece of work 👍 just a few comments but nothing crazy

**Review by Olshansk — COMMENTED** [review]
@deblasis I'm still thinking through a couple of your comments and will follow up if I have a better idea. In the meantime, left some responses and addressed some of the changes.

File: utility/block.go (hunk 13 above)
Comment (Olshansk) [reply utility/block.go:60]: I get it. I have a slightly bigger fan and it's also spinning. Let's see if they're spinning at the same frequency.

1. I rename all instances of `transaction` to `tx` in the utility module
2. The comments still use the full word for readability purposes
3.  I keep the protobuf named `Transaction` for clarity purposes

File: utility/types/message.go (hunk 3 above)
Comment (Olshansk) [reply utility/types/message.go]: Also added the following comment in the interface

```go
	// TECHDEBT: This is necessary for us to be able to apply some `proto` specific transformations
	//           on types conforming to this interface
	codec.CodecType
```

**Review by Olshansk — COMMENTED** [review]
@deblasis PTAL (please take another look).

Left some replies in the resolved comments as well as kept a few ones open. Will also update the chanelogs.

File: utility/block.go (hunk 13 above)
Comment (Olshansk) [reply utility/block.go]: > Not a big deal of course but historically, we invested time and thought in these kinds of situations

Fwiw, I did invest a lot (15? mins) of time thinking about it yesterday. If the decision doesn't feel "obvious", let's make it so.

**Reasoning for keeping the `type` called `Transaction`**:
- We refer to types in docs & diagrams
- We refer to types in presentations
- Types (and interfaces) are usually the "starting point" when you are trying to understand a codebase
- The type will define the state commitment (the hash), and there fully being explicit & long is helpful

Reasoning for keeping interfaces & vars `tx`:
- We use this keyword A LOT throughout the codebase
- It adds a lot of reading (longer code lines) and increases the cognitive load on the reader (experienced or not)
- I don't think it adds much value when it's repeated so often
- Being able to reference `someTx` in comments and conversations instead of `someTransaction` will add a lot bit by bit over time

Reasoning for keeping `transaction` in the comments:
- This one I'm open to going either way and feel less strongly about
- These comments will be read less often and by less experienced people than experienced devs
- Using full words "tells a story" of what's actually happening

Comment (deblasis) [reply utility/block.go]: Great points. You have my support

File: runtime/test_artifacts/defaults.go [hunk 17]
```diff
@@ -8,7 +8,7 @@ import (
 
 var (
 	DefaultChains              = []string{"0001"}
-	DefaultServiceURL          = ""
+	DefaultServiceURL          = "https://localhost.com"
```
Comment (deblasis) [inline runtime/test_artifacts/defaults.go:11]: Not sure about this value here. Below, in `ServiceUrlFormat` we define it as `(host):(port)`

File: shared/core/types/actor.go (hunk 5 above)
Comment (deblasis) [reply shared/core/types/actor.go:12]: TL;DR: Yep, nuke the 0.

I'd wrap in a func what you said: 
```go
	for actorTypeNum := range coreTypes.ActorType_name {
		if actorTypeNum == 0 {
			// skipping the zero value (unspecified actor type) since it's not relevant here
			continue
		}
		actorType := coreTypes.ActorType(actorTypeNum)

```
to return a []coreTypes.ActorType

File: shared/codec/codec.go (hunk 2 above)
Comment (deblasis) [reply shared/codec/codec.go]: As far as I understand you are trying to centralize the type that we can encode/decode. Right?

In that case I guess all the methods should use that and not `proto.Message`. Something like: 

```go
package codec

import (
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
)

// DOCUMENT: Some of the types through the code based (e.g. Transaction) are defined via serializable
// files (e.g. .proto files) and add functionality defined in go on top of it (e.g. ITransaction). This
// a very implicit design pattern that may be non-obvious to new developers.
type CodecType proto.Message

// CONSIDERATION: Use generics in place of `proto.Message` in the interface below so
//
//	every caller does not need to do in place casting.
type Codec interface {
	Marshal(CodecType) ([]byte, error)
	Unmarshal([]byte, CodecType) error
	ToAny(CodecType) (*anypb.Any, error)
	FromAny(*anypb.Any) (CodecType, error)
	Clone(CodecType) CodecType
}

var _ Codec = &protoCodec{}

// IMPROVE: Need to define a type similar to `type ProtoAny anypb.Any` so that we are
//
//	referencing protobuf specific types (e.g. anypb.Any) anywhere in the codebase.
type protoCodec struct{}

func GetCodec() Codec {
	return &protoCodec{}
}

// IMPROVE: If/when we move Pocket's `Error` type into a separate package, we can return `ErrProtoMarshal` here
func (p *protoCodec) Marshal(msg CodecType) ([]byte, error) {
	return proto.MarshalOptions{Deterministic: true}.Marshal(msg)
}

func (p *protoCodec) Unmarshal(bz []byte, msg CodecType) error {
	return proto.Unmarshal(bz, msg)
}

func (p *protoCodec) ToAny(msg CodecType) (*anypb.Any, error) {
	return anypb.New(msg)
}

func (p *protoCodec) FromAny(any *anypb.Any) (CodecType, error) {
	return anypb.UnmarshalNew(any, proto.UnmarshalOptions{})
}

func (p *protoCodec) Clone(msg CodecType) CodecType {
	return proto.Clone(msg)
}
```
but at that point I would take care of `FromAny` and `ToAny` since are protobuf specific.

Also, question: are we planning to support different encodings apart from proto? If not, this sounds a bit overkill and prematurely optimized. I'd just restore the `proto.Message` embedding in the `Message`

Thoughts?

**Review by deblasis — APPROVED** [review]
@Olshansk I guess the pending convos are:

- https://github.com/pokt-network/pocket/pull/503#discussion_r1106255286
- https://github.com/pokt-network/pocket/pull/503#discussion_r1106098366
- https://github.com/pokt-network/pocket/pull/503#discussion_r1106267045

They are mostly questions and nits. I feel confident to just approve at this point. The main points have been already addressed I reckon

File: runtime/test_artifacts/defaults.go (hunk 17 above)
Comment (Olshansk) [reply runtime/test_artifacts/defaults.go:11]: Reverting.

File: utility/block.go (hunk 13 above)
Comment (Olshansk) [reply utility/block.go]: Captured it here: https://github.com/pokt-network/pocket/discussions/511

Comment (Olshansk) [reply utility/block.go]: Also adding this comment:

```
// `Transaction` is used to name the type for clarity & verbosity, but `tx` is used method signatures
// and variable names to be concise. https://github.com/pokt-network/pocket/pull/503
```

File: shared/core/types/actor.go (hunk 5 above)
Comment (Olshansk) [reply shared/core/types/actor.go]: Went with what I believ eis the best solution here: [`6733ec4` (#503)](https://github.com/pokt-network/pocket/pull/503/commits/6733ec493c6afaf17fa094afd7599277ca0bb16c)

No redundant type, just use proto directly w/ an if everywhere - very golangy
